overlord/hookstate/ctlcmd/fde_setup.go

```diff
@@ -0,0 +1,160 @@
+
+Alternatively the hook could reply with:
+$ echo '{"error":"hardware-unsupported"}' | snapctl fde-setup-result
+
+And then it is called again with a request to do the initial key setup:
+$ snapctl fde-setup-request
+{"op":"initial-setup", "key": "key-to-seal", "key-name":"key-for-ubuntu-data"}
+$ echo "$sealed_key" | snapctl fde-setup-result
+`)
+
+func init() {
+	addCommand("fde-setup-request", shortFdeSetupRequestHelp, longFdeSetupRequestHelp, func() command { return &fdeSetupRequestCommand{} })
+}
+
+type fdeSetupJSON struct {
+	// XXX: make "op" a type: "features", "initial-setup", "update" ?
+	Op string `json:"op"`
+
+	Key     []byte `json:"key,omitempty"`
+	KeyName string `json:"key-name,omitempty"`
+
+	// Model related fields, this will be set to follow the
+	// secboot:SnapModel interface.
+	//
+	// XXX: do we need this to be a list? i.e. multiple models?
+	Model map[string]string `json:"model,omitempty"`
+
+	// TODO: provide LoadChains, KernelCmdline etc to support full
+	//       tpm sealing
+}
+
+func (c *fdeSetupRequestCommand) Execute(args []string) error {
+	context := c.context()
+	if context == nil {
+		return fmt.Errorf("cannot  without a context")
+	}
+	context.Lock()
+	defer context.Unlock()
+
+	var js fdeSetupJSON
+	if err := context.Get("fde-op", &js.Op); err != nil {
+		return fmt.Errorf("cannot get fde op from context: %v", err)
+	}
+	// Op is either "initial-setup" or "features"
+	switch js.Op {
+	case "features":
+		// nothing
+	case "initial-setup":
+		if err := context.Get("fde-key", &js.Key); err != nil {
+			return fmt.Errorf("cannot get fde key from context: %v", err)
+		}
+		if err := context.Get("fde-key-name", &js.KeyName); err != nil {
+			return fmt.Errorf("cannot get fde key name from context: %v", err)
+		}
+		if err := context.Get("model", &js.Model); err != nil {
+			return fmt.Errorf("cannot get model from context: %v", err)
+		}
+	default:
+		return fmt.Errorf("unknown fde-setup-request op %q", js.Op)
+
+	}
+
+	bytes, err := json.Marshal(js)
+	if err != nil {
+		return fmt.Errorf("cannot json print fde key: %v", err)
+	}
+	c.printf("%s\n", string(bytes))
+
+	return nil
+}
+
+type fdeSetupResultCommand struct {
+	baseCommand
+}
+
+var shortFdeSetupResultHelp = i18n.G("Set result for full disk encryption")
+var longFdeSetupResultHelp = i18n.G(`
+The fde-setup-result command reads the result data for a fde-setup hook
+from stdin.
+
+E.g.
+When the fde-setup hook is called with "op":"features:
+$ echo "[]" | snapctl fde-setup-result
+
+Or when the fde-setup hook is called with "op":"initial-setup":
+$ echo "sealed-key" | snapctl fde-setup-result
+`)
+
+func init() {
+	addCommand("fde-setup-result", shortFdeSetupResultHelp, longFdeSetupResultHelp, func() command { return &fdeSetupResultCommand{} })
+}
+
+func (c *fdeSetupResultCommand) Execute(args []string) error {
+	context := c.context()
+	if context == nil {
+		return fmt.Errorf("cannot  without a context")
+	}
+	context.Lock()
+	defer context.Unlock()
+
```

Comment:
should we check first that there was a fde-setup-op on the context  to again give a good error if this is the wrong kind of hook?

Reply:
this isn't addressed